File: csharp/exercise/exer06a-blocking-queue.cs
/*
 * BLOCKING QUEUE IMPLEMENTATION
 * Version A: Synchronous queues
 */
using System;
using System.Threading;



class Exer06A : IRunnable
{
    public void run()
    {
        var queue = new MySynchronousQueue<string>();
        new Thread(() => producer(queue)).Start();
        new Thread(() => consumer(queue)).Start();
    }


    private void producer(MySynchronousQueue<string> queue)
    {
        string[] arr = { "lorem", "ipsum", "dolor" };

        foreach (var data in arr)
        {
            Console.WriteLine($"Producer: {data}");
            queue.put(data);
            Console.WriteLine($"Producer: {data}\t\t\t[done]");
        }
    }


    private void consumer(MySynchronousQueue<string> queue)
    {
        string data;
        Thread.Sleep(5000);

        for (int i = 0; i < 3; ++i)
        {
            data = queue.take();
            Console.WriteLine($"\tConsumer: {data}");
        }
    }


    class MySynchronousQueue<T>
    {
        private Semaphore semPut = new Semaphore(1, int.MaxValue);
        private Semaphore semTake = new Semaphore(0, int.MaxValue);
        private T element = default;


        public void put(T value)
        {
            semPut.WaitOne();
            element = value;
            semTake.Release();
        }


        public T take()
        {
            semTake.WaitOne();

            T result = element;
            element = default;

            semPut.Release();

            return result;
        }
    }
}

File: python/exer06a_blocking_queue.py
'''
BLOCKING QUEUE IMPLEMENTATION
Version A: Synchronous queues
'''

import time
import threading



class SynchronousQueue:
    def __init__(self):
        self.__sem_put = threading.Semaphore(1)
        self.__sem_take = threading.Semaphore(0)
        self.__element = None

    def put(self, value):
        self.__sem_put.acquire()
        self.__element = value
        self.__sem_take.release()

    def take(self):
        self.__sem_take.acquire()
        result = self.__element
        self.__sem_put.release()
        return result



def producer(syncq: SynchronousQueue):
    arr = [ 'lorem', 'ipsum', 'dolor' ]

    for data in arr:
        print(f'Producer: {data}')
        syncq.put(data)
        print(f'Producer: {data} \t\t\t[done]')



def consumer(syncq: SynchronousQueue):
    time.sleep(5)

    for _ in range(3):
        value = syncq.take()
        print(f'\tConsumer: {value}')



syncqueue = SynchronousQueue()
threading.Thread(target=producer, args=(syncqueue,)).start()
threading.Thread(target=consumer, args=(syncqueue,)).start()
